src/types/memoryview.rs

```diff
@@ -0,0 +1,123 @@
+use std::marker::PhantomData;
+use std::os::raw::c_char;
+
+use crate::{AsPyPointer, PyAny, PyResult, Python};
+use crate::buffer::{Element, PyBuffer};
+use crate::ffi::{self, Py_ssize_t};
+use crate::types::PySequence;
+
+/// Represents a Python `memoryview`.
+#[repr(transparent)]
+pub struct PyMemoryView<T>(PyAny, PhantomData<T>);
+
+pyobject_native_type_core!(PyMemoryView<T>, ffi::PyMemoryView_Type, #checkfunction=ffi::PyMemoryView_Check; T);
+
+impl PyMemoryView<u8> {
+    pub fn new<'p>(py: Python<'p>, memory: &mut [u8], is_writable: bool) -> &'p PyMemoryView<u8> {
+        let ptr = memory.as_mut_ptr() as *mut c_char;
+        let len = memory.len() as Py_ssize_t;
+        let flags = if is_writable {
+            ffi::PyBUF_WRITE
+        } else {
+            ffi::PyBUF_READ
+        };
+
+
+        unsafe { py.from_owned_ptr(ffi::PyMemoryView_FromMemory(ptr, len, flags)) }
+    }
+
+    /// Creates a new Python memoryview from a raw pointer and length.
+    ///
+    /// Panics if out of memory.
+    ///
+    /// # Safety
+    ///
+    /// This function dereferences the raw pointer `ptr` as the
+    /// leading pointer of a slice of length `len`. [As with
+    /// `std::slice::from_raw_parts`, this is
+    /// unsafe](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html#safety).
+    pub unsafe fn from_ptr(py: Python<'_>, ptr: *const u8, len: usize, is_writable: bool) -> &PyMemoryView<u8> {
+        let flags = if is_writable {
+            ffi::PyBUF_WRITE
+        }
+        else {
+            ffi::PyBUF_READ
+        };
+
+        py.from_owned_ptr(ffi::PyMemoryView_FromMemory(
+            ptr as *mut _,
+            len as Py_ssize_t,
+            flags
+        ))
+    }
+}
+
+impl<T: Element> PyMemoryView<T> {
+
+    /// Creates a new Python `memoryview` object from another Python object that
+    /// implements the buffer protocol.
+    pub fn from<'p, I>(py: Python<'p>, src: &'p I) -> PyResult<&'p PyMemoryView<T>>
```

Comment:
I might be tempted to just have `pub fn from(&PyAny) -> PyResult<&PyMemoryView>` to simplify the signature a bit.